### PR TITLE
added abstract.py

### PR DESCRIPTION
**Pull request by pkijania**

### CODE REVIEW

File: abstract.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        self._price = price
+        self._date = date
+        self._ema_long = ema_long
+        self._ema_short = ema_short
+        self._macd = macd
+        self._rsi = rsi
+        self._adx = adx
+
+    @property
+    def price(self):
```
Comment (dkijania) [inline abstract.py]: You should add property for each field in class

File: abstract.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+from abc import ABC, abstractmethod
+
+class DataManager(ABC):
```
Comment (dkijania) [inline abstract.py]: I think we should also add abstract method for clear like

```
	@abstractmethod
	def clean(self): 
	    pass
```

File: main.py [hunk 3]
```diff
@@ -52,11 +52,13 @@ def main(data_file, break_time):
         strategy.order()
         buy, sell = strategy.get_order()
 
-        # Put all the information in a 'bitcoin_price.csv' file
-        csv_file.append_data_file(price, date, ema_long, ema_short, macd, rsi, adx)
-
-        # Put all the information in a postgres data base
-        postgres.transfer_data(price, date, ema_long, ema_short, macd, rsi, adx)
+        # Put all teh information to csv file or postgres data base
+        if data_file.startswith("postgres://"):
```
Comment (dkijania) [inline main.py]: i think we discussed about better name. Not 
`@click.option('--data_file', help = 'Path to datafile.')` but

`@click.option('--data_source', help = 'Data source for persisting data. Can be either path to csv or connection string to postgres database')`

File: abstract.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    @abstractmethod
+    def insert(self, model):
```
Comment (dkijania) [inline abstract.py]: Good. Now the next step is to modify PostgresDatabase class in postgres.py and Csv class in csv.py so they will inherit from DataManager

File: abstract.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from abc import ABC, abstractmethod
```
Comment (dkijania) [inline abstract.py]: I think the better place for this method id in `data` folder

File: main.py [hunk 6]
```diff
@@ -10,14 +10,18 @@
...
-def main(data_file, break_time):
-    # Clear all data from 'bitcoin_price.csv'
-    csv_file = CsvFile(data_file)
-    csv_file.remove_data_file()
-
-    postgres = PostgresDataBase()
+def main(data_source, break_time):
+    if data_source.startswith("postgres://"):
+        # Clear all data from 'postgres data base'
+        postgres = PostgresDataBase()
```
Comment (dkijania) [inline main.py]: in order to benefit fully from abstract class we can write something like:
```
    if data_source_startswith("postgres://"):
	datamanager = PostgresDataBase()
    else:
        datamanager = CsvFile(data_source)
        
        
        datamanager.clean()
        
        ...
        datamanager.insert(...)
```

as a benefit you don't need second if in line 60.

File: data/postgres.py [hunk 7]
```diff
@@ -1,15 +1,19 @@
+from data.abstract import DataManager
 import psycopg2
 
-class PostgresDataBase:
+class PostgresDataBase(DataManager):
     def __init__(self):
         print("Transfering data to postgres data base")
+        self.conn = psycopg2.connect(host = "localhost", dbname = "postgres", user = "postgres", password = "toster1111", port = 5432)
```
Comment (dkijania) [inline data/postgres.py]: We should not have hardcoded values like this In our code. In main.py we are accepting connection string which should have format like:

postgres://postgres:toster1111@localhost:5432/crazy_tr8

I think psycopg2 does not provide constructor which can accept connection string but separate parameters like in your code. We need to have then a prestep which will unwrap those values from connection string:

https://stackoverflow.com/questions/70939969/psycopg2-connect-to-postgresql-database-using-a-connection-string

Warning ⚠️:

Format in above link is different (they have localhost as schema, we don't, so we need to change logic when unwrapping parameters.

File: data/postgres.py [hunk 8]
```diff
@@ -1,15 +1,19 @@
...
+        self.conn.set_session(autocommit=True)
 
-    def transfer_data(self, price, date, ema_long, ema_short, macd, rsi, adx):
-        conn = psycopg2.connect(host = "localhost", dbname = "postgres", user = "postgres", password = "toster1111", port = 5432)
-        cur = conn.cursor()
+    def insert(self, price, date, ema_long, ema_short, macd, rsi, adx):
         insert_values = 'INSERT INTO crazy_tr8 (price, date, ema_long, ema_short, macd, rsi, adx) VALUES (%s, %s, %s, %s, %s, %s, %s)'
```
Comment (dkijania) [inline data/postgres.py]: Crazy_tr8 should not be a table but a schema.

Usually we should let the user create a schema and choose which user he/she wants to use. We should also require that the schema already has all the tables (like we do here). I think we should implement well established approach with dealing with database. Some responsibilities are on user side some on ours:

1. We should provide create_schema.sql which user can use to create database
2. User should setup database and user ans the schema on his/her own and provide connection string when using our app for example: (postgres://postgres:toster1111@localhost:5432/crazy_tr8)

To sum up the flow is:

1. We are committing create_schema.sql file to repo
2. User is creating database using for example:

psql -u postgres -c "create database crazy_tr8"

3. User run our create_schema.sql file on newly created dB:

psql -u postgres -d crazy_tr8 -f create_schema.sql

4. User starts app with -datasource postgres://postgres:toster1111@localhost:5432/crazy_tr8

**Review by dkijania — CHANGES_REQUESTED** [review]
